### ISSUE TITLE
Update facets of IDs

### ISSUE THREAD

**Issue by joncison** [bio-tools/biotoolsSchema#79]

**Comment by joncison** [comment]
and make CollectionID match

**Comment by matuskalas** [comment]
As there are more updates necessary for the IDs, I'm extending this issue and listing them here:

- [ ] Update maxlen facet on toolID to match what's currently supported in bio.tools (was the original title of this issue). Personally, I don't see a need for a formal maxlen constraint. What do others think? Or is for "anti-spam" purposes? I personally don't have anything against a maxlen constraint either, as long as it is a __round__ number (i.e. 32 or 64, but not 50 or 100 ;-D). 16 is too short, so 32 is my vote. And making a BiotoolsID simpleType in the XSD, used for tools, collections, etc.

- [ ] Make all IDs subtype of xs:NCName, with a matching pattern. Otherwise they can't be used in RDF/XML!! A good pattern for _bio.tools_ that matches NCName would be something like `[_a-z][_\-.0-9a-z]{,31}` or ` [_a-zA-Z][_\-.0-9a-zA-Z]{,31}` (see next one). Possibly with some non-English characters such as `é`, but perhaps rather not.

- [ ] I don't mind if capital letters are allowed (they're pretty), but both the __http resolution__ and the __uniqueness__ must work case-insensitively! _E.g._ http://bio.tools/pigNaLc should resolve|http-rewrite|redirect to http://bio.tools/PignalC . Is this how it works in _bio.tools_, @ekry ? If yes, awesome!!! If not, it should! ;-)

- [ ] Having _/tool_ or _/collection_ in the URI is undesirable for many reasons:
    - decreased usability
    - less of a "cool URI"
    - unnecessarily verbose
    - IDs of tools and collection should anyway be unique and distinct
    - Who should remember whether EMBOSS is http://bio.tools/tool/EMBOSS or http://bio.tools/collection/EMBOSS ?!?
    - We anyway shouldn't allow both http://bio.tools/tool/EMBOSS and http://bio.tools/collection/EMBOSS !
    - http://bio.tools already includes "tool", why repeat it?!? It feels very stuttering, just like http://bio.tools/bio/tools/tool 👎

**Comment by joncison** [comment]
- ``maxlen`` facet : I think we'll know a sensible value once a big clean-up of existing IDs is complete (@hans & I are busy with this).  The idea is it should be big enough to allow all sensible names (but not crazy big)

- there is currently ``biotooldIType`` simpleType (used in versionID and toolID) and ``biotoolsUrlType`` (used for ``relation->biotoolsId``) which includes valid bio.tools domains.  Probably the later type is not needed (just use ID w/o URL bit ?)

- bio.tools Tool Card URLs do resolve in case-insensitive way

- the plan indeed is to support *bio.tools/tooName*  (and probably retain bio.tools/t/toolName, bio.tools/tool/toolName - namespaces may be useful in future if we start e..g to provide information about collections etc.)

**Comment by joncison** [comment]
*UPDATE* cc @matuskalas @hansioan 

The new model for identifiers is:
![capture](https://user-images.githubusercontent.com/1506863/32404526-d2b23336-c148-11e7-8bca-c2cd0d852306.PNG)

`value` has no `maxlen` facet.  As all the bio.tools toolIDs are now manually verified anyway (Hans please see https://github.com/bio-tools/biotoolsregistry/issues/279) this should be OK.  It's `xs:token` with appropriate regex's (neither `xs:QName` or xs:NCName` would work).

The `maxLen` facet on `biotoolsIdType` (used in `relation->biotoolsId` is removed).

bio.tools URLs are now of the simpler form and, yes _e.g._  https://bio.tools/SiGnalp resolves

**Comment by matuskalas** [comment]
Could you please tick off the boxes of https://github.com/bio-tools/biotoolsSchema/issues/79#issuecomment-306793420 according to which are done?

To my understanding:
* the 1st is done by removing the `maxLength`, _i.e._ no lenght constraint. But is this also compatible with the functionality of _bio.tools_?
* the last, 5th point is done for tools, which is really excellent, hooray! The XSD of `biotoolsUrlType` needs to be updated accordingly.
* the 3rd point **is ABSOLUTELY CRUCIAL and should be fixed.**

**Comment by joncison** [comment]
I've ticked all the boxes.  bio.tools will be updated - beginning next week - to what will become biotoolsSchema 3.0.0.

Except for use of NCName, which I'm not sure about.   NCName preclude the use of ":", and without the prefixes (which include ":") it's not possible to disambiguate instances of different types of identifier value.  That would subsequently force us to make typing (via `identifier->type`) of the supplied identifier values mandatory at registration time (whether via UI or API).  

Rather, I think we should treat all identifiers consistently as CURIES (https://www.w3.org/TR/2010/NOTE-curie-20101216/), more specifically as having syntax `prefix:value` and let any application logic drop the prefix, or indeed map the CURIE to something resolvable, as required.  

The implication here we need a "canonical" prefix for bio.tools.  Logically, this would be "biotools" giving _e.g. `biotools:signalp` which is trivially mapable to the (resolvable) https://bio.tools/signalp.  

cc @hansioan @matuskalas @ekry

**Comment by joncison** [comment]
Going down this route, this is what the regex looks like for `identifier->value`:
![capture](https://user-images.githubusercontent.com/1506863/32660001-f688b026-c618-11e7-9705-26ee950aa5ff.PNG)

with examples:
![capture](https://user-images.githubusercontent.com/1506863/32660018-11f1517e-c619-11e7-9f45-06e0a065abed.PNG)

and in this case it (neatly) allow us to not have to mandate explicit typing of the supplied value separately, thus:
![capture](https://user-images.githubusercontent.com/1506863/32660127-787a4a2c-c619-11e7-9668-01aa30878fa8.PNG)

**Comment by joncison** [comment]
@matuskalas @hansioan @ekry - for now I will assume this is a reasonable fix, and my next commit will close this issue (can always be reopened as needed).   Like Matus, I'm not 100% sure that the bio.tools ID should be under `identifer` or in specific in it's own element.  Unless someone shouts, I leave it nested, as shown above.

**Comment by matuskalas** [comment]
This is all really awesome, epic developments around the bio.tools IDs (now and recently)! 👍 

Please check and accept (_i.e._ merge) the pull request #93, which fixes also the last standing point of https://github.com/bio-tools/biotoolsSchema/issues/79#issuecomment-306793420. After the merge, it will automatically and ultimately close this issue once for good! ;-)

**Comment by matuskalas** [comment]
The only thing left is https://github.com/bio-tools/biotoolsSchema/issues/79#issuecomment-306027592, _i.e._ making the collection ID (and URI) match. As there are some further considerations that need to be (under)taken, I created a separate new issue for that: #94.

**Comment by joncison** [comment]
@matuskalas, thanks for above!  

Now that `biotoolsID` is in it's own element and gives the "vanilla ID" e.g. "signalp"  (with `biotoolsCURIE` element for prefix'ed IDs e.g. "biotools:signalp") then we can change the type of  `biotoolsIdType` simpleType (as used by `biotoolsID`) from xs:anyURI to NCName.  My next commit will do this, and close this issue.

**Comment by joncison** [comment]
I ticked the 3rd box: xs:NCName is specified for ``biotoolsIdType`` simpleType thus ``biotoolsID`` wherever it occurs.

**Comment by joncison** [comment]
Following the discussion in https://github.com/bio-tools/biotoolsregistry/issues/284 biotoolsIdType (thus biotoolsID) is being reverted to xs:token with regex `[_\-.0-9a-zA-Z]*`.

**but**

next commit will add support for xs:NCName-compatible biotoolIDs to be specified in `otherID->value`, *i.e.* if it turns out we need such IDs, they can (will) be supported as alternative IDs.  I hope this is a reasonable compromise and avoids the refactoring (https://github.com/bio-tools/biotoolsregistry/issues/285, https://github.com/bio-tools/biotoolsregistry/issues/284) and other complications that go with this.
